velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,383 @@
+        std::shared_ptr<const ITypedExpr> vCallTypeInput =
+            vCallTypeInputs.at(i);
+        // TODO
+        //  need to judge according the names in the expr, and then set them to
+        //  the if or then or else expr can debug to find when process project
+        //  node
+      }
+    } else if (vCallTypeExprFunName == exec::kSwitch) {
+      ::substrait::Expression_SwitchExpression* sFun =
+          sExpr->mutable_switch_expression();
+      // TODO
+    } else {
+      ::substrait::Expression_ScalarFunction* sFun =
+          sExpr->mutable_scalar_function();
+      // TODO need to change yaml file to register functin, now is dummy.
+      // the substrait communcity have changed many in this part...
+      uint32_t sFunId =
+          v2SFuncConvertor_.registerSubstraitFunction(vCallTypeExprFunName);
+      sFun->set_function_reference(sFunId);
+
+      for (auto& vArg : vCallTypeInputs) {
+        ::substrait::Expression* sArg = sFun->add_args();
+        toSubstraitExpr(vArg, vPreNodeOutPut, sArg);
+      }
+      ::substrait::Type* sFunType = sFun->mutable_output_type();
+      v2STypeConvertor_.toSubstraitType(vExprType, sFunType);
+      return;
+    }
+  }
+  if (auto vFieldExpr =
+          std::dynamic_pointer_cast<const FieldAccessTypedExpr>(vExpr)) {
+    // kSelection
+    const std::shared_ptr<const Type> vExprType = vFieldExpr->type();
+    std::string vExprName = vFieldExpr->name();
+
+    ::substrait::Expression_ReferenceSegment_StructField* sDirectStruct =
+        sExpr->mutable_selection()
+            ->mutable_direct_reference()
+            ->mutable_struct_field();
+
+    std::vector<std::string> vPreNodeColNames = vPreNodeOutPut->names();
+    std::vector<std::shared_ptr<const velox::Type>> vPreNodeColTypes =
+        vPreNodeOutPut->children();
+    int64_t vPreNodeColNums = vPreNodeColNames.size();
+    int64_t sCurrentColId = -1;
+
+    VELOX_CHECK_EQ(vPreNodeColNums, vPreNodeColTypes.size());
+
+    for (int64_t i = 0; i < vPreNodeColNums; i++) {
+      if (vPreNodeColNames[i] == vExprName &&
+          vPreNodeColTypes[i] == vExprType) {
+        sCurrentColId = i;
+        break;
+      }
+    }
+
+    if (sCurrentColId == -1) {
+      sCurrentColId = vPreNodeColNums + 1;
+    }
+    sDirectStruct->set_field(sCurrentColId);
+
+    return;
+  }
+  if (auto vCastExpr = std::dynamic_pointer_cast<const CastTypedExpr>(vExpr)) {
+    std::vector<std::shared_ptr<const ITypedExpr>> vCastTypeInputs =
+        vCastExpr->inputs();
+    ::substrait::Expression_Cast* sCastExpr = sExpr->mutable_cast();
+    v2STypeConvertor_.toSubstraitType(
+        vCastExpr->type(), sCastExpr->mutable_type());
+
+    for (auto& vArg : vCastTypeInputs) {
+      toSubstraitExpr(vArg, vPreNodeOutPut, sCastExpr->mutable_input());
+    }
+    return;
+
+  } else {
+    VELOX_UNSUPPORTED("Unsupport Expr '{}' in Substrait", vExpr->toString());
+  }
+}
+
+void VeloxToSubstraitExprConvertor::toSubstraitLiteral(
```

Comment:
see my comment below; can we have this function to take the constant expr directly, and dispatch based on whether variant or vector is available wihin?